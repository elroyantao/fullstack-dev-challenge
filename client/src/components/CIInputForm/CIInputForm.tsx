import { VStack, FormControl, FormLabel } from '@chakra-ui/react'
import React, { useEffect, useState } from 'react'
import { IProjectionRequest } from '../../api/compoundInterest/calculateProjection'
import NumberSlider from '../../uicomponents/NumberSlider'

type Props = {
    onChange: (data: IProjectionRequest) => void
}

const CIInputFrom: React.FC<Props> = ({ onChange }: Props) => {
    const [principal, setPrincipal] = useState<number>(0)
    const [monthlyDeposit, setMonthlyDeposit] = useState<number>(0)
    const [interestRate, setInterestRate] = useState<number>(0)

    useEffect(() => {
        if (principal && monthlyDeposit && interestRate) {
            onChange({
                principal: principal,
                monthlyDeposit: monthlyDeposit,
                interestRate: interestRate,
            })
        }
    }, [principal, monthlyDeposit, interestRate, onChange])

    return (
        <VStack height="100%" justify="space-between">
            <FormControl id="principal">
                <FormLabel fontSize="sm" fontWeight="normal">
                    Initial deposit
                </FormLabel>
                <NumberSlider
                    value={principal}
                    onChange={setPrincipal}
                    max={1000000}
                    maxSlider={100000}
                />
            </FormControl>

            <FormControl id="interest">
                <FormLabel fontSize="sm" fontWeight="normal">
                    Interest Rate
                </FormLabel>
                <NumberSlider
                    value={interestRate}
                    onChange={setInterestRate}
                    max={20}
                    maxSlider={20}
                    step={0.01}
                />
            </FormControl>

            <FormControl id="monthlyDeposit">
                <FormLabel fontSize="sm" fontWeight="normal">
                    Monthly deposit
                </FormLabel>
                <NumberSlider
                    value={monthlyDeposit}
                    onChange={setMonthlyDeposit}
                    max={10000}
                    maxSlider={10000}
                />
            </FormControl>
        </VStack>
    )
}

export default CIInputFrom
