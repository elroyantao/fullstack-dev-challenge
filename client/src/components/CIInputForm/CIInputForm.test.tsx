import React from 'react'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import { act } from 'react-dom/test-utils'

import CIInputForm from './CIInputForm'

const onChangeMock = jest.fn()

beforeEach(() => {
    onChangeMock.mockClear()
})

describe('<CIInputFrom />', () => {
    describe('@renders', () => {
        it('default state', () => {
            const wrapper = shallow(<CIInputForm onChange={onChangeMock} />)
            expect(toJson(wrapper)).toMatchSnapshot()
            expect(onChangeMock).toHaveBeenCalledTimes(0)
            expect(wrapper.find('NumberSlider').at(0).prop('value')).toBe(0)
            expect(wrapper.find('NumberSlider').at(1).prop('value')).toBe(0)
            expect(wrapper.find('NumberSlider').at(2).prop('value')).toBe(0)
        })
    })

    const waitForComponentToPaint = async (wrapper: any) => {
        await act(async () => {
            await new Promise(setImmediate)
            wrapper.update()
        })
    }

    describe('@side-effect', () => {
        it('should called onchange with the request params when all the values are entered', async () => {
            const wrapper = mount(<CIInputForm onChange={onChangeMock} />)

            expect(wrapper.find('NumberSlider').at(0).prop('value')).toBe(0)
            expect(wrapper.find('NumberSlider').at(1).prop('value')).toBe(0)
            expect(wrapper.find('NumberSlider').at(2).prop('value')).toBe(0)

            act(() => {
                const { onChange } = wrapper.find('NumberSlider').at(0).props()
                onChange(1000)
            })
            await waitForComponentToPaint(wrapper)
            expect(wrapper.find('NumberSlider').at(0).prop('value')).toBe(1000)
            expect(onChangeMock).toHaveBeenCalledTimes(0)

            act(() => {
                const { onChange } = wrapper.find('NumberSlider').at(1).props()
                onChange(2)
            })
            await waitForComponentToPaint(wrapper)
            expect(wrapper.find('NumberSlider').at(1).prop('value')).toBe(2)
            expect(onChangeMock).toHaveBeenCalledTimes(0)

            act(() => {
                const { onChange } = wrapper.find('NumberSlider').at(2).props()
                onChange(100)
            })
            await waitForComponentToPaint(wrapper)
            expect(wrapper.find('NumberSlider').at(2).prop('value')).toBe(100)

            expect(onChangeMock).toHaveBeenCalledTimes(1)
            expect(onChangeMock).toHaveBeenLastCalledWith({
                principal: 1000,
                interestRate: 2,
                monthlyDeposit: 100,
            })
        })
    })
})
