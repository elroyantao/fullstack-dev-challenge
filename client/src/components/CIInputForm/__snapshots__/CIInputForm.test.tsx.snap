// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CIInputFrom /> @renders default state 1`] = `
<VStack
  height="100%"
  justify="space-between"
>
  <FormControl
    id="principal"
  >
    <FormLabel
      fontSize="sm"
      fontWeight="normal"
    >
      Initial deposit
    </FormLabel>
    <NumberSlider
      max={1000000}
      maxSlider={100000}
      onChange={[Function]}
      value={0}
    />
  </FormControl>
  <FormControl
    id="interest"
  >
    <FormLabel
      fontSize="sm"
      fontWeight="normal"
    >
      Interest Rate
    </FormLabel>
    <NumberSlider
      max={20}
      maxSlider={20}
      onChange={[Function]}
      step={0.01}
      value={0}
    />
  </FormControl>
  <FormControl
    id="monthlyDeposit"
  >
    <FormLabel
      fontSize="sm"
      fontWeight="normal"
    >
      Monthly deposit
    </FormLabel>
    <NumberSlider
      max={10000}
      maxSlider={10000}
      onChange={[Function]}
      value={0}
    />
  </FormControl>
</VStack>
`;
